composer/index.ts

```diff
@@ -0,0 +1,168 @@
+    ) {}
+}
+
+class Composition {
+    constructor(public global: Global, public runs: Run[]) {}
+}
+
+async function main() {
+    sqlite3.verbose();
+
+    // Call sqlite to process the csv resource files and generate a database.
+    // We call the sqlite process instead of doing it here cause
+    // the dot commands are interpreted by the sqlite cli tool not sqlite itself,
+    // and it is a lot faster parsing the csv's.
+    const { stdout, stderr } = await execFile("sqlite3", [
+        DB,
+        ".mode csv",
+        ".import transports.csv transports",
+        ".import muxers.csv muxers",
+    ]);
+    if (stderr != "") {
+        throw new Error(`Could not parse csv resources: ${stderr}`);
+    }
+
+    const db = await open({
+        filename: DB,
+        driver: sqlite3.Database,
+    });
+
+    // Generate the testing combinations by SELECT'ing from both transports
+    // and muxers tables the distinct combinations where the transport and the muxer
+    // of the different libp2p implementations match.
+    const queryResults =
+        await db.all(`SELECT DISTINCT a.id as id1, b.id as id2, a.transport, ma.muxer
+                     FROM transports a, transports b, muxers ma, muxers mb
+                     WHERE a.id != b.id
+                     AND a.transport == b.transport
+                     AND a.id == ma.id
+                     AND b.id == mb.id
+                     AND ma.muxer == mb.muxer;`);
+    await db.close();
+
+    let global = new Global(
+        "multidimensional-testing",
+        "multidimensional",
+        argv.total_instances
+    );
+    let composition = new Composition(global, []);
+
+    for (let row of queryResults) {
+        // Instance count is hardcoded to 1 for now.
+        let instance = new Instance(1);
+
+        let build_args1 = new BuildArgs(
+            row.transport,
+            row.muxer,
+            row.id1,
+            argv.gitRef,
+            argv.gitTarget
+        );
+        let build_config1 = new BuildConfig(build_args1);
+        let group1 = new Group(row.id1, BUILDER, instance, build_config1);
+
+        let build_args2 = new BuildArgs(
+            row.transport,
+            row.muxer,
+            row.id2,
+            argv.gitRef,
+            argv.gitTarget
+        );
+        let build_config2 = new BuildConfig(build_args2);
+        let group2 = new Group(row.id2, BUILDER, instance, build_config2);
+
+        let run = new Run(
+            `${row.id1} x ${row.id2} x ${row.transport} x ${row.muxer}`,
+            [group1, group2]
+        );
+
+        composition.runs.push(run);
+    }
+
+    // Write the TOML file and remove the database file to avoid corrupting
+    // future runs.
+    await writeFile("composition.toml", toml.stringify(composition as any));
+    await unlink(DB);
+}
+
+main()
+    .then(() => process.exit(0))
+    .catch((err) => {
+        console.error(err);
+        process.exit(1);
+    });
```

Comment:
(creating here just to get a thread)

> pass all the test parameters (transport, muxer, protocol) when launching the container the work becomes a lot easier for the composer and we could also combine it with https://github.com/testground/testground/issues/1522.
Does testground currently support passing env vars when launching containers @laurentsenta?

Yes, the idiomatic solution is to use the run parameters,
there is an example of this in the ping test:
https://github.com/libp2p/test-plans/blob/abce1ee4832d04cc9854266df71378b440a68a5e/ping/go/main.go#L69-L71

You'd pass the parameters in the `runs` directly or in the `runs.group` field:
https://github.com/libp2p/test-plans/blob/3fccfb27bbaba204437b93a73d8d718f4b372442/composer/demo/composition-example.toml#L79-L80